File: src/main/java/com/rj/ecommerce_email_service/messaging/producer/EmailMessageProducer.java
package com.rj.ecommerce_email_service.messaging.producer;

import com.rj.ecommerce_email_service.messaging.common.AbstractMessageProducer;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.stereotype.Component;

@Component
@Slf4j
public class EmailMessageProducer extends AbstractMessageProducer {


    public EmailMessageProducer(RabbitTemplate rabbitTemplate) {
        super(rabbitTemplate);
    }

    public void sendEmail(String exchange, String routingKey, Object message, String correlationId) {
        sendMessage(exchange, routingKey, message, correlationId);
    }
}

File: backend/src/main/kotlin/com/rj/ecommerce_backend/messaging/email/producer/EmailMessageProducer.kt
package com.rj.ecommerce_backend.messaging.email.producer

import com.rj.ecommerce.api.shared.messaging.email.EmailRequest
import com.rj.ecommerce_backend.messaging.common.producer.AbstractMessageProducer
import com.rj.ecommerce_backend.messaging.config.RabbitMQProperties
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.stereotype.Component

@Component
class EmailMessageProducer(
    rabbitTemplate: RabbitTemplate,
    private val rabbitMQProperties: RabbitMQProperties
) : AbstractMessageProducer(rabbitTemplate) {

    /**
     * Sends a generic, composed email request to the email exchange.
     */
    fun send(request: EmailRequest<*>) {
        val emailConfig = rabbitMQProperties.email
        sendMessage(
            exchange = emailConfig.exchange,
            routingKey = emailConfig.routingKey,
            message = request,
            correlationId = request.envelope.correlationId
        )
    }
}